File: templates/guardrails/fitness/pureza-dominio.mjs
#!/usr/bin/env node
/**
 * Fitness function: la capa de dominio es pura (equivalente TypeScript de
 * `pureza-dominio.py`).
 *
 * POR QUÉ EXISTE
 * ──────────────
 * La misma regla que en Python: "el dominio no importa infraestructura" es
 * una propiedad ESTRUCTURAL del proyecto, no una regla del lenguaje — ESLint
 * o Biome no tienen forma de saber que `src/dominio/` es especial sin una
 * regla a medida tan grande como este archivo mismo.
 *
 * QUÉ MIDE (y qué NO)
 * ────────────────────
 * Recorre cada archivo bajo `RUTA_DOMINIO` y extrae sus imports (estático y
 * dinámico) y sus `require(...)`. Si alguno resuelve a un módulo de
 * `IMPORTS_PROHIBIDOS`, es una violación — con archivo y línea.
 *
 * NO mide que el dominio sea correcto — eso lo prueban los tests de negocio.
 * Mide que no dependa de lo que el diseño prohíbe que dependa.
 *
 * ANÁLISIS DE IMPORTS, NO UN AST COMPLETO DE TYPESCRIPT — Y POR QUÉ
 * ────────────────────────────────────────────────────────────────
 * Un AST completo de TypeScript necesitaría el compilador de TypeScript
 * como dependencia (`typescript`), y este kit mantiene sus guardrails sin
 * dependencias externas (sólo `node:*`). La alternativa —parsear imports
 * con expresiones regulares sobre el texto, después de sacar comentarios—
 * es más barata y alcanza para lo que esta fitness function necesita: los
 * imports son, casi siempre, sintaxis de una sola línea al principio del
 * archivo. La limitación real: un import reconstruido con concatenación de
 * strings o con un identificador dinámico (`import(variable)`) no se
 * detecta — es el mismo tipo de límite que `check-config-duplicada.mjs`
 * documenta para su propio barrido de texto: costo de VISIBILIDAD, no
 * imposibilidad. Si tu proyecto necesita cerrar ese caso, subí este checker
 * a un plugin de `ts-morph` o del compilador de TypeScript.
 *
 * POR QUÉ EL LINTER / CI NORMAL NO LO DETECTA
 * ─────────────────────────────────────────────
 * Porque "esta carpeta no puede importar esta otra cosa" es una regla del
 * PROYECTO, no del lenguaje — ESLint tiene un plugin (`no-restricted-
 * imports`) que se le acerca, pero configurarlo para una lista dinámica y
 * con el mensaje explicativo que este archivo da requiere casi el mismo
 * código. La ventaja de un checker propio es que el mensaje de error puede
 * decir EXACTAMENTE por qué la regla existe.
 *
 * CONFIGURACIÓN
 * ─────────────
 * Completá `RUTA_DOMINIO` con la carpeta real del dominio del proyecto, y
 * ajustá `IMPORTS_PROHIBIDOS`.
 */

import { readdirSync, readFileSync, statSync } from "node:fs";
import { dirname, join, resolve } from "node:path";
import { fileURLToPath } from "node:url";

const ROOT = resolve(dirname(fileURLToPath(import.meta.url)), "../..");

// ── CONFIGURACIÓN: ajustá esto al proyecto real ─────────────────────────────

/** Carpeta del dominio a verificar, relativa a la raíz del repo. */
const RUTA_DOMINIO = "src/dominio";

/**
 * Módulos (o prefijos de módulo) que el dominio NO puede importar. Un
 * prefijo matchea con `/` — ej. `"react"` NO matchea `"react-query-algo"`,
 * pero si querés bloquear un scope entero usá `"@nestjs/"` (con la barra).
 */
const IMPORTS_PROHIBIDOS = [
  "fs",
  "node:fs",
  "http",
  "node:http",
  "axios",
  "express",
  "next",
  "react",
  "pg",
  "drizzle-orm",
  "@prisma/client",
];

/** Extensiones de archivo que se recorren. */
const EXTENSIONES = new Set([".ts", ".tsx", ".js", ".jsx", ".mjs", ".cjs"]);

// ── el checker ───────────────────────────────────────────────────────────────

/** Saca comentarios de línea y de bloque, sin tocar el contenido de strings
 * (aproximado: no distingue un `//` dentro de un template literal — es la
 * misma limitación documentada arriba, aceptada a propósito). */
function sinComentarios(codigo) {
  return codigo
    .replace(/\/\*[\s\S]*?\*\//g, (bloque) => bloque.replace(/[^\n]/g, " "))
    .replace(/(^|[^:])\/\/.*$/gm, "$1");
}

/**
 * Extrae los especificadores de módulo importados de un archivo: `import …
 * from "x"`, `import "x"`, `export … from "x"`, `require("x")`,
 * `import("x")` dinámico.
 *
 * @returns {{ especificador: string, linea: number }[]}
 */
function extraerImports(codigo) {
  const limpio = sinComentarios(codigo);
  const lineas = limpio.split("\n");
  const resultado = [];

  const patrones = [
    /\bfrom\s+["']([^"']+)["']/,
    /\brequire\(\s*["']([^"']+)["']\s*\)/,
    /\bimport\(\s*["']([^"']+)["']\s*\)/,
    /^\s*import\s+["']([^"']+)["']\s*;?\s*$/,
  ];

  lineas.forEach((linea, indice) => {
    for (const patron of patrones) {
      const match = linea.match(patron);
      if (match) {
        resultado.push({ especificador: match[1], linea: indice + 1 });
      }
    }
  });

  return resultado;
}

/** ¿Este especificador cae bajo algún prohibido de la lista? */
function esProhibido(especificador) {
  return IMPORTS_PROHIBIDOS.some(
    (prohibido) => especificador === prohibido || especificador.startsWith(`${prohibido}/`)
  );
}

/** Barrido recursivo de archivos con extensión relevante. */
function archivosDeDominio(dir, encontrados = []) {
  let entradas;
  try {
    entradas = readdirSync(dir);
  } catch {
    return encontrados;
  }
  for (const entrada of entradas) {
    if (entrada === "node_modules" || entrada.startsWith(".")) continue;
    const completo = join(dir, entrada);
    const info = statSync(completo);
    if (info.isDirectory()) {
      archivosDeDominio(completo, encontrados);
    } else if (EXTENSIONES.has(entrada.slice(entrada.lastIndexOf(".")))) {
      encontrados.push(completo);
    }
  }
  return encontrados;
}

/**
 * @param {string} raiz ruta absoluta de la carpeta de dominio
 * @returns {{ archivo: string, linea: number, motivo: string }[]}
 */
export function verificarPurezaDominio(raiz) {
  const violaciones = [];
  for (const archivo of archivosDeDominio(raiz)) {
    const codigo = readFileSync(archivo, "utf8");
    for (const { especificador, linea } of extraerImports(codigo)) {
      if (esProhibido(especificador)) {
        violaciones.push({
          archivo,
          linea,
          motivo: `import prohibido en el dominio: '${especificador}'`,
        });
      }
    }
  }
  return violaciones;
}

// ── veredicto ────────────────────────────────────────────────────────────────

const raizDominio = resolve(ROOT, RUTA_DOMINIO);

let existe = true;
try {
  statSync(raizDominio);
} catch {
  existe = false;
}

if (!existe) {
  console.error(`\n✗ No existe ${raizDominio}. Ajustá RUTA_DOMINIO en este script.\n`);
  process.exit(1);
}

const violaciones = verificarPurezaDominio(raizDominio);

if (violaciones.length > 0) {
  console.error(`\nEl dominio (${RUTA_DOMINIO}) no es puro — ${violaciones.length} violación(es):\n`);
  for (const v of violaciones) {
    console.error(`✖ ${v.archivo.replace(`${ROOT}/`, "")}:${v.linea}  ${v.motivo}`);
  }
  console.error(
    "\nPara arreglarlo: sacá la dependencia de infraestructura del dominio — invertí la\n" +
      "dependencia (el dominio define una interfaz/puerto, la infraestructura la implementa\n" +
      "y se inyecta) en vez de importar el módulo concreto desde adentro del dominio.\n"
  );
  process.exit(1);
}

console.info(`✓ El dominio (${RUTA_DOMINIO}) es puro — sin imports prohibidos`);
process.exit(0);

File: templates/guardrails/fitness/pureza-dominio.py
#!/usr/bin/env python3
"""Fitness function: la capa de dominio es pura.

POR QUE EXISTE
--------------
Un linter de Python no tiene forma de expresar "prohibido importar `requests`
(o cualquier cosa de infraestructura/IO) dentro de `dominio/`" sin convertirse
en un plugin a medida. La regla es real -- el dominio que depende de una
libreria HTTP, de un ORM o de un reloj de pared (`datetime.now()`) deja de ser
testeable sin red, sin base de datos, sin mockear el tiempo -- pero es
ESTRUCTURAL, no sintactica, y por eso el linter estandar no la ve.

Esta fitness function la convierte en un CHEQUEO EJECUTABLE: recorre el AST
real de cada archivo bajo la carpeta de dominio y reporta cada violacion con
archivo, linea y motivo. No es un grep -- un grep se enganya con un string, un
comentario o un import dentro de un docstring; el AST no.

QUE MIDE (y que NO)
--------------------
Cuatro clases de violacion, todas configurables en las constantes de abajo:
  1. imports prohibidos (IMPORTS_PROHIBIDOS)          -- ej. "requests", "os"
  2. llamadas prohibidas (LLAMADAS_PROHIBIDAS)        -- ej. "open()" (I/O)
  3. atributos prohibidos (ATRIBUTOS_PROHIBIDOS)      -- ej. ".now" (reloj implicito)
  4. literales/tipos prohibidos (TIPOS_LITERALES_PROHIBIDOS) -- ej. float en dinero

NO mide que el dominio sea CORRECTO -- eso lo prueban los tests de negocio.
Mide que el dominio sea PURO: nada de IO, nada de infraestructura, nada de
relojes o generadores de aleatoriedad implicitos.

POR QUE EL LINTER NORMAL NO LO DETECTA
----------------------------------------
Porque "importar `os` esta prohibido" no es una regla del lenguaje Python --
es una regla de ESTE proyecto, sobre ESTA carpeta. mypy/ruff no tienen forma
de saber que "dominio/" es especial sin una configuracion a medida tan grande
como este archivo mismo.

CONFIGURACION
-------------
Completa RUTA_DOMINIO con la carpeta real del dominio del proyecto, y ajusta
las cuatro constantes de prohibiciones. Corre como script (`python
fitness/pureza_dominio.py`) o importa `verificar_pureza_dominio()` desde un
test de pytest (ver `fitness/README.md`, seccion "como se escribe una": un
checker sin un test que confirme que detecta la violacion es una afirmacion
de fe).
"""

from __future__ import annotations

import ast
import sys
from dataclasses import dataclass
from pathlib import Path

# ── CONFIGURACION: ajusta esto al proyecto real ─────────────────────────────

# Carpeta del dominio a verificar, relativa a la raiz del repo.
RUTA_DOMINIO = Path("src/dominio")

# Modulos que el dominio NO puede importar. `open` y `datetime.now` no son
# imports (son builtin/atributo) y se verifican aparte, mas abajo.
IMPORTS_PROHIBIDOS: frozenset[str] = frozenset(
    {"fastapi", "flask", "django", "pydantic", "httpx", "requests", "sqlalchemy", "csv", "os"}
)

# Llamadas de I/O directo prohibidas dentro del dominio.
LLAMADAS_PROHIBIDAS: frozenset[str] = frozenset({"open", "print", "input"})

# Atributos que representan un reloj o una fuente de aleatoriedad implicita
# -- ambos son I/O disfrazado: hacen que el mismo codigo devuelva resultados
# distintos en corridas distintas, lo que rompe la property mas importante
# del dominio (determinismo, testeable sin mockear tiempo/azar).
ATRIBUTOS_PROHIBIDOS: frozenset[str] = frozenset({"now", "utcnow", "random"})

# Tipos que no deberian aparecer como literal en el dominio. El ejemplo
# clasico es `float` en calculos monetarios: el redondeo binario de punto
# flotante produce resultados distintos segun el orden de las operaciones,
# y una factura mal redondeada no falla ruidosamente -- cobra mal en
# silencio. Dejalo vacio (frozenset()) si no aplica a tu dominio.
TIPOS_LITERALES_PROHIBIDOS: frozenset[str] = frozenset({"float"})

# ── el checker ───────────────────────────────────────────────────────────────


@dataclass(frozen=True, slots=True)
class ViolacionPureza:
    archivo: Path
    linea: int
    motivo: str

    def __str__(self) -> str:
        return f"{self.archivo}:{self.linea}: {self.motivo}"


def verificar_pureza_dominio(raiz: Path) -> list[ViolacionPureza]:
    """Recorre `raiz` buscando `.py` y devuelve TODAS las violaciones de pureza."""
    violaciones: list[ViolacionPureza] = []
    for archivo in sorted(raiz.rglob("*.py")):
        codigo = archivo.read_text(encoding="utf-8")
        try:
            arbol = ast.parse(codigo, filename=str(archivo))
        except SyntaxError as error:
            violaciones.append(ViolacionPureza(archivo, error.lineno or 0, f"no parsea: {error}"))
            continue
        for nodo in ast.walk(arbol):
            violaciones.extend(_violaciones_de_nodo(nodo, archivo))
    return violaciones


def _violaciones_de_nodo(nodo: ast.AST, archivo: Path) -> list[ViolacionPureza]:
    if (
        isinstance(nodo, ast.Constant)
        and isinstance(nodo.value, float)
        and "float" in TIPOS_LITERALES_PROHIBIDOS
    ):
        return [ViolacionPureza(archivo, nodo.lineno, "literal float prohibido en el dominio")]
    if (
        isinstance(nodo, ast.Name)
        and nodo.id in TIPOS_LITERALES_PROHIBIDOS
    ):
        return [ViolacionPureza(archivo, nodo.lineno, f"uso del tipo '{nodo.id}' prohibido en el dominio")]
    if _es_llamada_prohibida(nodo):
        assert isinstance(nodo, ast.Call) and isinstance(nodo.func, ast.Name)
        motivo = f"llamada a '{nodo.func.id}()' prohibida en el dominio (I/O)"
        return [ViolacionPureza(archivo, nodo.lineno, motivo)]
    if isinstance(nodo, ast.Attribute) and nodo.attr in ATRIBUTOS_PROHIBIDOS:
        motivo = f"acceso a '.{nodo.attr}' prohibido en el dominio (reloj/azar implicito)"
        return [ViolacionPureza(archivo, nodo.lineno, motivo)]
    if isinstance(nodo, (ast.Import, ast.ImportFrom)):
        return _violaciones_de_import(nodo, archivo)
    return []


def _es_llamada_prohibida(nodo: ast.AST) -> bool:
    return (
        isinstance(nodo, ast.Call)
        and isinstance(nodo.func, ast.Name)
        and nodo.func.id in LLAMADAS_PROHIBIDAS
    )


def _violaciones_de_import(
    nodo: ast.Import | ast.ImportFrom, archivo: Path
) -> list[ViolacionPureza]:
    if isinstance(nodo, ast.Import):
        modulos = [alias.name.split(".")[0] for alias in nodo.names]
    else:
        modulos = [(nodo.module or "").split(".")[0]]
    return [
        ViolacionPureza(archivo, nodo.lineno, f"import prohibido en el dominio: '{modulo}'")
        for modulo in modulos
        if modulo in IMPORTS_PROHIBIDOS
    ]


def main() -> int:
    raiz_repo = Path(__file__).resolve().parents[2]
    raiz_dominio = raiz_repo / RUTA_DOMINIO

    if not raiz_dominio.exists():
        print(f"✗ No existe {raiz_dominio}. Ajustá RUTA_DOMINIO en este script.", file=sys.stderr)
        return 1

    violaciones = verificar_pureza_dominio(raiz_dominio)

    if violaciones:
        print(f"\nEl dominio ({RUTA_DOMINIO}) no es puro — {len(violaciones)} violación(es):\n", file=sys.stderr)
        for v in violaciones:
            print(f"✖ {v}", file=sys.stderr)
        print(
            "\nPara arreglarlo: sacá la dependencia de infraestructura del dominio — invertí la\n"
            "dependencia (el dominio define una interfaz, la infraestructura la implementa) en\n"
            "vez de importar la librería concreta desde adentro del dominio.\n",
            file=sys.stderr,
        )
        return 1

    print(f"✓ El dominio ({RUTA_DOMINIO}) es puro — sin imports/llamadas/tipos prohibidos")
    return 0


if __name__ == "__main__":
    sys.exit(main())
